sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/repository/metric/LowMemoryMetricsRepository.java

```diff
@@ -0,0 +1,313 @@
+            for (Entry<String, LinkedHashMap<Long, LowMetricEntity>> resourceMetrics : resourceMap.entrySet()) {
+                for (Entry<Long, LowMetricEntity> metrics : resourceMetrics.getValue().entrySet()) {
+                    if (metrics.getKey() < minTimeMs) {
+                        continue;
+                    }
+                    LowMetricEntity newEntity = metrics.getValue();
+                    if (resourceCount.containsKey(resourceMetrics.getKey())) {
+                        LowMetricEntity oldEntity = resourceCount.get(resourceMetrics.getKey());
+                        oldEntity.addPassQps(newEntity.getPassQps());
+                        oldEntity.addRtAndSuccessQps(newEntity.getRt(), newEntity.getSuccessQps());
+                        oldEntity.addBlockQps(newEntity.getBlockQps());
+                        oldEntity.addExceptionQps(newEntity.getExceptionQps());
+                        oldEntity.addCount(1);
+                    } else {
+                        resourceCount.put(resourceMetrics.getKey(), LowMetricEntity.copyOf(newEntity));
+                    }
+                }
+            }
+            // Order by last minute b_qps DESC.
+            return resourceCount.entrySet()
+                    .stream()
+                    .sorted((o1, o2) -> {
+                        LowMetricEntity e1 = o1.getValue();
+                        LowMetricEntity e2 = o2.getValue();
+                        int t = e2.getBlockQps() < e1.getBlockQps() ? -1 : ((e2.getBlockQps() == e1.getBlockQps()) ? 0 : 1);
+                        if (t != 0) {
+                            return t;
+                        }
+                        return e2.getPassQps() < e1.getPassQps() ? -1 : ((e2.getPassQps() == e1.getPassQps()) ? 0 : 1);
+                    })
+                    .map(Entry::getKey)
+                    .collect(Collectors.toList());
+        } finally {
+            readWriteLock.readLock().unlock();
+        }
+    }
+}
+
+
+class LowMetricEntity {
+    private long timestamp;
+    private long gmtCreate;
+
+    private int successQps;
+    private int exceptionQps;
+    private int passQps;
+    private int blockQps;
+
+    /**
+     * summary rt of all success exit qps.
+     */
+    private double rt;
+
+    /**
+     * 本次聚合的总条数
+     */
+    private int count;
+
+
+    public synchronized void addPassQps(int passQps) {
+        this.passQps += passQps;
+    }
+
+    public synchronized void addBlockQps(int blockQps) {
+        this.blockQps += blockQps;
+    }
+
+    public synchronized void addExceptionQps(int exceptionQps) {
+        this.exceptionQps += exceptionQps;
+    }
+
+    public synchronized void addCount(int count) {
+        this.count += count;
+    }
+
+    public synchronized void addRtAndSuccessQps(double avgRt, int successQps) {
+        this.rt += avgRt * successQps;
+        this.successQps += successQps;
+    }
+
+    public static LowMetricEntity copyOf(LowMetricEntity oldEntity) {
```

Comment:
if `oldEntity` is null ? and should we add a lock during the function?

Reply:
i'm sorry but i can't understand why should we add a lock here.  Is  the lock meaning that we need to check whether the param is null? thanks for your help

Reply:
> if `oldEntity` is null ? and should we add a lock during the function?

i'm sorry but i can't understand why should we add a lock here. Is the lock meaning that we need to check whether the param is null? thanks for your help

Reply:
emmm, Checking  whether the `oldEntity` is null avoids null of point exception.and I see you use `synchronize` in  add  operation  like `addCount` function ,so I am thinking the `copyOf` function is the same as  `addCount` function

Reply:
understand. I think there is no need to add a lock for these methods.  `LowMetricEntity.toLowMetricEntity` is called with a writeLock; 
`LowMetricEntity.toMetricEntity` is called with a readLock; 
`LowMetricEntity.copyOf` is called with a readLock; 

the `synchronize` methods like `addCount`, is for a copy LowMetricEntity instance.

Reply:
> emmm, Checking whether the `oldEntity` is null avoids null of point exception.and I see you use `synchronize` in add operation like `addCount` function ,so I am thinking is
> the `copyOf` function the same as `addCount` function? If this progress is concurrently safe,my advice is unnecessary

maybe the oldEntity impossible to null, need i still do a null check?

Reply:
> emmm, Checking whether the `oldEntity` is null avoids null of point exception.and I see you use `synchronize` in add operation like `addCount` function ,so I am thinking is
> the `copyOf` function the same as `addCount` function? If this progress is concurrently safe,my advice is unnecessary

Thanks! understand. I think there is no need to add a lock for these methods. `LowMetricEntity.toLowMetricEntity` is called with a writeLock;
`LowMetricEntity.toMetricEntity` is called with a readLock;
`LowMetricEntity.copyOf` is called with a readLock;
the `synchronize` methods like `addCount`, is for a copy LowMetricEntity instance.

maybe the oldEntity impossible to null, need i still do a null check?

Reply:
> 
> 
> > emmm, Checking whether the `oldEntity` is null avoids null of point exception.and I see you use `synchronize` in add operation like `addCount` function ,so I am thinking is
> > the `copyOf` function the same as `addCount` function? If this progress is concurrently safe,my advice is unnecessary
> 
> Thanks! understand. I think there is no need to add a lock for these methods. `LowMetricEntity.toLowMetricEntity` is called with a writeLock;
> `LowMetricEntity.toMetricEntity` is called with a readLock;
> `LowMetricEntity.copyOf` is called with a readLock;
> the `synchronize` methods like `addCount`, is for a copy LowMetricEntity instance.
> 
> maybe the oldEntity impossible to null, need i still do a null check?

if you are sure ,it is not necessary